File: csharp/robot-name/RobotName.cs
using System;
using System.Text;
using System.Collections.Generic;

public class Robot
{
    public static HashSet<string> usedNames = new HashSet<string>();

    public Robot()
    {
         this.Name = NameGenerator();
    }
    public string Name
    {
        get; private set;
    }

    private string NameGenerator()
    {
        StringBuilder sb = new StringBuilder();
        var random = new Random();

        sb.Append(Convert.ToChar(random.Next(65,91)));
        sb.Append(Convert.ToChar(random.Next(65, 91)));
        sb.Append(Convert.ToString(random.Next(100, 1000)));
        string newName = sb.ToString();
        if (usedNames.Contains(newName)) { newName = NameGenerator(); }
        usedNames.Add(newName);

        return newName;
    }

    public void Reset()
    {
        this.Name = NameGenerator();
    }
}

File: javascript/robot-name/robot-name.js
// This is only a SKELETON file for the 'Robot Name' exercise. It's been
// provided as a convenience to get your started writing code faster.
let usedNames = {};

export class Robot { 
    constructor(){
        this.reset();// first time boot up is basically a reset
    }
    
    get name(){
        return this._name;
    }
    reset(){// gives the robot a new name
        this._name = nameGenerator();
    }
}

function nameGenerator() { // strings together the required number of each character and redoes it if it isnt unique
    let newName = '';
    do{
        newName = randomCap() + randomCap() + randomNum() + randomNum() + randomNum();
    } while(usedNames[newName]);
    usedNames[newName] = true;
    return newName;
}

function randomCap(){ //picks a random cap character
    return String.fromCharCode(Math.floor((Math.random()*26) + 65));
}

function randomNum(){// picks a random number
    return Math.floor((Math.random()*10));
}

Robot.releaseNames = () => usedNames = {}; //this is required by Exercism and it resets the usedNames list.
